File: apps/web/lib/api/problem.ts
/** RFC 9457 problem details, as the API returns errors. */
type Problem = { detail?: string; title?: string };

function isProblem(value: unknown): value is Problem {
  return typeof value === "object" && value !== null && ("detail" in value || "title" in value);
}

/**
 * The user-facing message of a failed API call: the problem's detail (the
 * API keeps those safe to show), else its title, else the fallback.
 */
export function problemMessage(error: unknown, fallback = "Something went wrong. Please try again."): string {
  if (!isProblem(error)) return fallback;
  return error.detail?.trim() || error.title?.trim() || fallback;
}

File: apps/api/internal/transport/httpapi/problem.go
package httpapi

import (
	"context"
	"errors"
	"log/slog"
	"net/http"

	"github.com/danielgtaylor/huma/v2"

	"github.com/hamid-karimi/coachin/apps/api/internal/app/apperr"
	"github.com/hamid-karimi/coachin/apps/api/internal/app/auth"
)

var authStatus = map[auth.Kind]int{
	auth.Invalid:      http.StatusBadRequest,
	auth.Unauthorized: http.StatusUnauthorized,
	auth.Conflict:     http.StatusConflict,
}

var appStatus = map[apperr.Kind]int{
	apperr.Invalid:   http.StatusBadRequest,
	apperr.NotFound:  http.StatusNotFound,
	apperr.Forbidden: http.StatusForbidden,
	apperr.Conflict:  http.StatusConflict,
}

// toProblem turns a use-case error into an HTTP problem: expected failures
// keep their user-facing message; anything else is logged and hidden.
func toProblem(ctx context.Context, logger *slog.Logger, err error) error {
	var authErr *auth.Error
	if errors.As(err, &authErr) {
		return huma.NewError(authStatus[authErr.Kind], authErr.Message)
	}
	var appErr *apperr.Error
	if errors.As(err, &appErr) {
		return huma.NewError(appStatus[appErr.Kind], appErr.Message)
	}
	logger.ErrorContext(ctx, "request failed", "error", err)
	return huma.Error500InternalServerError("Something went wrong. Please try again.")
}
